src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,347 @@
+        }
+
+        /// <summary>
+        /// For tuple literals, we just return the element.
+        /// For expressions with tuple type, we access `Item{i}`.
+        /// </summary>
+        private BoundExpression GetTuplePart(BoundExpression tuple, int i)
+        {
+            // Example:
+            // (1, 2) == (1, 2);
+            if (IsTupleExpression(tuple.Kind))
+            {
+                return ((BoundTupleLiteral)tuple).Arguments[i];
+            }
+
+            Debug.Assert(tuple.Type.IsTupleType);
+
+            // Example:
+            // t == GetTuple();
+            // t == ((byte, byte)) (1, 2);
+            // t == ((short, short))((int, int))(1L, 2L);
+            return MakeTupleFieldAccessAndReportUseSiteDiagnostics(tuple, tuple.Syntax, tuple.Type.TupleElements[i]);
+        }
+
+        /// <summary>
+        /// Produce an element-wise comparison and logic to ensure the result is a bool type.
+        ///
+        /// If an element-wise comparison doesn't return bool, then:
+        /// - if it is dynamic, we'll do `!(comparisonResult.false)` or `comparisonResult.true`
+        /// - if it implicitly converts to bool, we'll just do the conversion
+        /// - otherwise, we'll do `!(comparisonResult.false)` or `comparisonResult.true` (as we'd do for `if` or `while`)
+        /// </summary>
+        private BoundExpression RewriteTupleSingleOperator(TupleBinaryOperatorInfo.Single single,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType, BinaryOperatorKind operatorKind)
+        {
+            if (single.Kind.IsDynamic())
+            {
+                // Produce
+                // !((left == right).op_false)
+                // (left != right).op_true
+
+                BoundExpression dynamicResult = _dynamicFactory.MakeDynamicBinaryOperator(single.Kind, left, right, isCompoundAssignment: false, _compilation.DynamicType).ToExpression();
+                if (operatorKind == BinaryOperatorKind.Equal)
+                {
+                    return _factory.Not(MakeUnaryOperator(UnaryOperatorKind.DynamicFalse, left.Syntax, method: null, dynamicResult, boolType));
+                }
+                else
+                {
+                    return MakeUnaryOperator(UnaryOperatorKind.DynamicTrue, left.Syntax, method: null, dynamicResult, boolType);
+                }
+            }
+
+            if (left.IsLiteralNull() && right.IsLiteralNull())
+            {
+                // For `null == null` this is special-cased during initial binding
+                return new BoundLiteral(left.Syntax, ConstantValue.Create(operatorKind == BinaryOperatorKind.Equal), boolType);
+            }
+
+            // PROTOTYPE(tuple-equality) checked
+            // We leave the null literal in nullable-null conversions unconverted because MakeBinaryOperator has special rules for it
+            bool isNullableNullConversion = ((operatorKind & BinaryOperatorKind.NullableNull) == 0);
```

Comment:
>bool isNullableNullConversion = ((operatorKind & BinaryOperatorKind.NullableNull) == 0); [](start = 12, length = 88)

Is this logic correct? Should we be checking for null here? Consider to use OperandTypes helper here as well.